frontend/catalyst/pennylane_extensions.py

```diff
@@ -2446,3 +2448,46 @@ def _get_batch_size(args_flat, axes_flat, axis_size):
         )
 
     return batch_size
+
+
+def callback(func):
+    """Decorator that will correctly pass the signature as arguments to the callback
+    implementation.
+
+    At the moment, this is a stub.
+    We can also use here isTracing to only bind with the callback while it is tracing,
+    otherwise, call the function itself.
+    """
+    signature = inspect.signature(func)
+    retty = signature.return_annotation
+
+    @wraps(func)
+    def bind_callback(*args, **kwargs):
+        callback_implementation(func, retty, *args, **kwargs)
+
+    return bind_callback
+
+
+def callback_implementation(
+    cb: Callable[..., Any], _result_shape_dtypes: Any, *args: Any, **kwargs: Any
+):
+    """TODO: Attribution. I looked into
+    https://jax.readthedocs.io/en/latest/_modules/jax/_src/callback.html#pure_callback"""
+    flat_args, in_tree = tree_flatten((args, kwargs))
+
+    def _flat_callback(*flat_args):
+        """This function packages flat arguments back into the shapes expected by the function."""
+        _args, _kwargs = tree_unflatten(in_tree, flat_args)
+        assert not _args, "Args are not yet expected here."
+        assert not _kwargs, "Kwargs are not yet supported here."
```

Comment:
We need to bind the flat list of arguments because JAX can only lower tensors to MLIR (as opposed to arbitrary classes). We then unflatten when receiving them such that the callback can expect dictionaries or any flattable / unflattable type.